### ISSUE TITLE
backward() in Autograd Index Function is broken when masking with ByteTensor

### ISSUE THREAD

**Issue by bermanmaxim** [pytorch/pytorch#971]
Hello, I hit this possible bug, similar to #828: the gradients do not seem to back-propagate through a masked array, e.g.:

```python
a = Variable(torch.range(1, 3), requires_grad=True)
valid = torch.ByteTensor([1, 0, 1])
b = a[valid]
c = b.sum()
c.backward()
a.grad
```
returns a `[0, 0, 0]` tensor.
(I have this result with PyTorch compiled from commit 5e7f5db3328a07d54e9fdd91fc20254331f0f1e4.

**Comment by bermanmaxim** [comment]
Oh wait, maybe this is fixed in current master, I just notices I was using commit  a version compiled from aaf41c61a657e188c6021d577de30ae041560a32

**Comment by bermanmaxim** [comment]
Yes, just tested with last master, it seems to have been fixed, sorry!
